rt_diversity_filter.go

```diff
@@ -0,0 +1,101 @@
+package dht
+
+import (
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-kbucket/peerdiversity"
+	ma "github.com/multiformats/go-multiaddr"
+	"sync"
+)
+
+var _ peerdiversity.PeerIPGroupFilter = (*rtPeerIPGroupFilter)(nil)
+
+// locking is the responsibility of the caller
+type rtPeerIPGroupFilter struct {
+	mu sync.RWMutex
+	h  host.Host
+
+	maxPerCpl   int
+	maxForTable int
+
+	cplIpGroupCount   map[int]map[peerdiversity.PeerIPGroupKey]int
+	tableIpGroupCount map[peerdiversity.PeerIPGroupKey]int
+}
+
+// NewRTPeerDiversityFilter constructs the `PeerIPGroupFilter` that will be used to configure
+// the diversity filter for the Routing Table.
+// Please see the docs for `peerdiversity.PeerIPGroupFilter` AND `peerdiversity.Filter` for more details.
+func NewRTPeerDiversityFilter(h host.Host, maxPerCpl, maxForTable int) *rtPeerIPGroupFilter {
+	return &rtPeerIPGroupFilter{
+		h: h,
+
+		maxPerCpl:   maxPerCpl,
+		maxForTable: maxForTable,
+
+		cplIpGroupCount:   make(map[int]map[peerdiversity.PeerIPGroupKey]int),
+		tableIpGroupCount: make(map[peerdiversity.PeerIPGroupKey]int),
+	}
+
+}
+
+func (r *rtPeerIPGroupFilter) Allow(g peerdiversity.PeerGroupInfo) bool {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	key := g.IPGroupKey
+	cpl := g.Cpl
+
+	if r.tableIpGroupCount[key] >= r.maxForTable {
+
+		return false
+	}
+
+	c, ok := r.cplIpGroupCount[cpl]
+	allow := !ok || c[key] < r.maxPerCpl
+	return allow
+}
+
+func (r *rtPeerIPGroupFilter) Increment(g peerdiversity.PeerGroupInfo) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	key := g.IPGroupKey
+	cpl := g.Cpl
+
+	r.tableIpGroupCount[key] = r.tableIpGroupCount[key] + 1
+	_, ok := r.cplIpGroupCount[cpl]
+	if !ok {
+		r.cplIpGroupCount[cpl] = make(map[peerdiversity.PeerIPGroupKey]int)
+	}
+	r.cplIpGroupCount[cpl][key] = r.cplIpGroupCount[cpl][key] + 1
+}
+
+func (r *rtPeerIPGroupFilter) Decrement(g peerdiversity.PeerGroupInfo) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	key := g.IPGroupKey
+	cpl := g.Cpl
+
+	r.tableIpGroupCount[key] = r.tableIpGroupCount[key] - 1
+	if r.tableIpGroupCount[key] == 0 {
+		delete(r.tableIpGroupCount, key)
+	}
+
+	r.cplIpGroupCount[cpl][key] = r.cplIpGroupCount[cpl][key] - 1
+	if r.cplIpGroupCount[cpl][key] == 0 {
+		delete(r.cplIpGroupCount[cpl], key)
+	}
+	if len(r.cplIpGroupCount[cpl]) == 0 {
+		delete(r.cplIpGroupCount, cpl)
+	}
+}
+
+func (r *rtPeerIPGroupFilter) PeerAddresses(p peer.ID) []ma.Multiaddr {
+	cs := r.h.Network().ConnsToPeer(p)
```

Comment:
Are there any issues here around a DoS attack if A connects to B with many different addresses, then A takes up all those spots in B's routing table?

Similarly, what if A connects to be with addresses 1 which is valid, then connects with address 2 which is invalid. Then disconnects address 1? Can this attack be extended such that address 2 can be reused for many attacker peers?

Reply:
As discussed offline, we want to keep this simple for now.